### PR TITLE
RPM sensing from GPIO (PWM pin) pulses

### PR DESCRIPTION
**Pull request by MaEtUgR**
### Solved Problem
When wokring on helicopter rpm control I found that RPM sensing is "easily" possible but requires work. @bkueng provided the nice configurable capture part for that effort and I used the output directly in the heli-specific controller. But now I think we should enable generic RPM capturing for use in control, log, UI reporting.

### Solution
~As a draft, I added the hardcoded logic we used to process the RPM.
Note: This doesn't work well out of the box because it cannot detect a timeout if the interrupt is never called anymore.~
The solution in the meantime has:
- timeout
- outlier rejection
- pule per revolution configuration

Thanks for all the help from @oravla5, @sfuhrer and @bresch behind the scenes.

### Changelog Entry
```
Feature: Easy to use RPM sensing from GPIO pulses
```

### Test coverage
This logic was used on a heli to control RPM before porting it to latest main and moving the processing to the rpm capture driver. EDIT: and changing it completely again.

~We need to follow up with further testing, settling the interface, implementing the timeout... FYI @sfuhrer~
We did testing on an RC helicopter with high pulse per revolution count + noisy signal and a combustion engine with outliers.

### CODE REVIEW

**Comment by sfuhrer** [comment]
We need to get better with rejecting outliers:
image

**Comment by oravla5** [comment]
> We need to get better with rejecting outliers: image

@sfuhrer 
Do you by any chance still have this ulog?

**Comment by MaEtUgR** [comment]
I rebased on main, fixed the screw up with the unwanted, unused `RpmControl.hpp` file I added and was changed for quick tests instead of the real logic. And added all our latest changes we came up with @oravla5.

> We did testing on an RC helicopter with high pulse per revolution count + noisy signal and a combustion engine with outliers.

Here's the plot from the small RC heli spinning the motor without blades open loop:
grafik

File: src/drivers/rpm_capture/rpm_capture_params.c
Comment (oravla5) [inline src/drivers/rpm_capture/rpm_capture_params.c:37]: Before merging, shall we check if we could use other channels different from 5? If so, we should update the comment accordingly.

**Review by oravla5 — COMMENTED** [review]
LGTM once we check if we can do RPM sensing in other channels that are not 5.

File: src/drivers/rpm_capture/rpm_capture_params.c
Comment (sfuhrer) [reply src/drivers/rpm_capture/rpm_capture_params.c:37]: Yes definitely. I always used 5 so far because of that comment...

Comment (MaEtUgR) [reply src/drivers/rpm_capture/rpm_capture_params.c:37]: I quickly asked the ST experts and the interrupt should be independent of timers and configuring the pin for input interrupt should be possible on most pins but probably depends on the board. So the assumption is it should work also on the other FMU PWM pins.

**Comment by oravla5** [comment]
I confrimed that RPM capture works on other AUX GPIO pins. It's not available in the MAIN pins.

I also played around with the parameters of the alpha and medium filters:
- Increasing the median filter window makes the rpm estimate significantly smoother for the small heli noisy raw RPM, but also slower. A value of 7 seems to be a good compromise.
- Increasing the alpha filter factor does not have a significant effect.

Find attached the logs for:
- Alpha filter 0.5, median filter window 5
- Alpha filter 0.5, median filter window 11
- Alpha filter 0.5, median filter window 7
- Alpha filter 0.7, median filter window 7

alpha0.5_median5
alpha0.5_median11
alpha0.5_median7
alpha0.7_median7

[rpm_caputure_filter_tests.zip](https://github.com/user-attachments/files/18097699/rpm_caputure_filter_tests.zip)

File: src/drivers/rpm_capture/rpm_capture_params.c
Comment (oravla5) [reply src/drivers/rpm_capture/rpm_capture_params.c:37]: Confirmed :ok_hand:

File: src/drivers/rpm_capture/CMakeLists.txt
Comment (bresch) [inline src/drivers/rpm_capture/CMakeLists.txt:3]:
```suggestion
#   Copyright (c) 2024 PX4 Development Team. All rights reserved.
```

File: msg/Rpm.msg
Comment (bresch) [inline msg/Rpm.msg:4]:
```suggestion
float32 rpm_raw
```

File: msg/Rpm.msg
Comment (bresch) [inline msg/Rpm.msg:5]:
```suggestion
float32 estimated_accurancy_rpm
```

File: src/drivers/rpm_capture/RPMCapture.cpp
Comment (bresch) [inline src/drivers/rpm_capture/RPMCapture.cpp:62]: Could we have a constexpr instead of a magic number?

File: src/drivers/rpm_capture/RPMCapture.cpp
Comment (bresch) [inline src/drivers/rpm_capture/RPMCapture.cpp:69]: why 2070?

File: src/drivers/rpm_capture/RPMCapture.cpp
Comment (bresch) [inline src/drivers/rpm_capture/RPMCapture.cpp:80]: Is it too difficult to add it now?

File: src/drivers/rpm_capture/RPMCapture.cpp
Comment (bresch) [inline src/drivers/rpm_capture/RPMCapture.cpp:141]: add a constexpr for "RPM filter time constant" ?

File: src/drivers/rpm_capture/RPMCapture.hpp
Comment (bresch) [inline src/drivers/rpm_capture/RPMCapture.hpp:3]:
```suggestion
 *   Copyright (c) 2024 PX4 Development Team. All rights reserved.
```

**Review by bresch — CHANGES_REQUESTED** [review]
Just a bit of nitpicking, otherwise looks good

File: src/drivers/rpm_capture/rpm_capture_params.c
Comment (bkueng) [reply src/drivers/rpm_capture/rpm_capture_params.c:37]: Channel 5 has history reasons. This is now more flexible as any pin can be configured, so the comment does not apply anymore.

File: src/drivers/rpm_capture/CMakeLists.txt
Comment (MaEtUgR) [reply src/drivers/rpm_capture/CMakeLists.txt:3]: It was in all the new files 😄

File: src/drivers/rpm_capture/RPMCapture.cpp
Comment (MaEtUgR) [reply src/drivers/rpm_capture/RPMCapture.cpp:69]: It's defined as a function in this pr:
https://github.com/PX4/PX4-Autopilot/blob/bdec521b0a629d47d053a31d303f41913e355dc8/src/lib/mixer_module/output_functions.yaml#L64
I'm not sure if that's defined outside of the metadata 👀 otherwise I'll extend the comment.

File: src/drivers/rpm_capture/CMakeLists.txt
Comment (MaEtUgR) [reply src/drivers/rpm_capture/CMakeLists.txt]: a5f81a19f3f33c50aea23722798cb1ba76cf8a12

File: src/drivers/rpm_capture/RPMCapture.hpp
Comment (MaEtUgR) [reply src/drivers/rpm_capture/RPMCapture.hpp]: a5f81a19f3f33c50aea23722798cb1ba76cf8a12

File: msg/Rpm.msg
Comment (MaEtUgR) [reply msg/Rpm.msg]: a5f81a19f3f33c50aea23722798cb1ba76cf8a12

File: msg/Rpm.msg
Comment (MaEtUgR) [reply msg/Rpm.msg]: a5f81a19f3f33c50aea23722798cb1ba76cf8a12

File: src/drivers/rpm_capture/RPMCapture.cpp
Comment (MaEtUgR) [reply src/drivers/rpm_capture/RPMCapture.cpp]: 31269c77f2cf6bbb586f02a22f06cbb3281295db

File: src/drivers/rpm_capture/RPMCapture.cpp
Comment (MaEtUgR) [reply src/drivers/rpm_capture/RPMCapture.cpp]: a5f81a19f3f33c50aea23722798cb1ba76cf8a12

File: src/drivers/rpm_capture/RPMCapture.cpp
Comment (MaEtUgR) [reply src/drivers/rpm_capture/RPMCapture.cpp]: d5f09596283d0d4c7ef128b7655acc1ac65e9848

**Comment by MaEtUgR** [comment]
@bresch I rebased on main, addressed all your comments in new commits (not one each but they're referenced)
and added 375cd553b9b2014b5796a325676ec12b742307c7 on top because I saw that was only necessary before https://github.com/PX4/PX4-Autopilot/commit/15fece7e141ee8c3a3357ceb1f0e7b55f0566558#diff-21b8564105b699e09127da22888e7b35b845e53be24c30a6f921ea1f9e82f821L84.

**Review by bresch — DISMISSED** [review]

**Comment by MaEtUgR** [comment]
`v6xrt_allyes` 👀 
grafik

🙈 
grafik

**Comment by MaEtUgR** [comment]
`v6xrt_allyes` fixed by https://github.com/PX4/PX4-Autopilot/commit/de04cb1ac4053e60ac96c459b991fc857abbe9fc#diff-5cc1f83d05caa7e38df255ec8fb58eeac44362c55bca89ffcfe55de562b46c82R43

`holybro_kakutef7` fixed by https://github.com/PX4/PX4-Autopilot/pull/24126 (needs to be reviewed)

**Comment by hamishwillee** [comment]
@MaEtUgR Can we have docs/info for this?

Specifically
- What kind of RPM sensors are supported? Do we have any sensors we can list that we know to work
- Is a tachometer sensor like this supported: https://docs.px4.io/main/en/sensor/tachometers.html
- That doc ^^^ says **Currently PX4 just logs RPM data: it is not used for state estimation or control.**. What use cases are we currently covering for this now - looks like helicopter RPM control (whatever that is) and ICE engines. Anything else?
- Given it is supported, is there any special setup required?
  - That tachometer suggest I2C as per https://docs.px4.io/main/en/sensor/thunderfly_tachometer.html#hardware-setup - but looking at this it looks like you require AUX/PWM input - or is that an "As well" and as long as it gets into the `Rpm` message you don't care?
- What does this support in helicopters that is now undocumented. I'm thinking updates to [helicopter types](https://docs.px4.io/main/en/frames_helicopter/#helicopter-types) and [Config](https://docs.px4.io/main/en/config_heli/)

Anything else you think I am missing!

**Pull request merged**
